### PR TITLE
Repository transfer should be confirmed by the new owner before the transfer actually occurs

### LINKED ISSUE #4743: Transfer repo ownership should be confirmed by new owner
- Gitea version (or commit ref): b1ad573 
- Git version: any
- Operating system: any
- Database (use `[x]`): any
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (provide example URL) https://try.gitea.io/test/frame-test
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

User can easily frame another user. How to reproduce:
1. Create new repo.
2. Configure local git to use victim's name/email.
3. Add to this repo something illegal in victim's country (or something agressive about other group of people or just smething very shameful).
4. Transfer ownership of repo to victim.
5. Call the police (or press, victim's boss, victim's wife or anyone else who will hate victim for publishing such information).

From (2) there is no protection, because it's how git itself works (it allows unsigned commits by design).
But from (4) there should be some kind of protection. For example, on Github anyone cannot simply transfer repo ownership to anyone, new owner should confirm this transfer.

Between user/org, user should be able to transfer repo without confirmation if he has admin rights in this org. If transferring repo to another org, any org admin should be able to confirm this transfer. If transferring repo to another user, another user should confirm transfer. And actual transfer must be performed after confirmation. And if user has admin rights, he probably should be able to transfer repo without confirmation (as he implicitly has all privileges on server).

Upd: additionally, when transfer is pending and waiting for user confirmation, new repo owner should have temporary read access to repo if it's private - to make him able to investigate this repo content before accepting/rejecting transfer.

## Screenshot

How it looks on Github:

### PR DESCRIPTION
**Pull request by adelowo**
Fixes https://github.com/go-gitea/gitea/issues/4743

- [x] Don't transfer a repository immediately but ___put the repo in a pending state___. See screenshot below
- [x] Cancel a repository transfer. 
- [x] Send an email to the recipient informing him/her of the pending transfer.
- [x] Actually perform the transfer once the recipient accepts the "invite"
- [x] Transfer invitations can be rejected
- [x] Make a redirect after a successful transfer
- [x] Tests.

screenshot 2019-01-16 at 15 08 18

### CODE REVIEW

File: models/repo.go [hunk 1]
```diff
@@ -1444,120 +1444,6 @@ func RepoPath(userName, repoName string) string {
 	return filepath.Join(UserPath(userName), strings.ToLower(repoName)+".git")
 }
 
-// TransferOwnership transfers all corresponding setting from old user to new one.
```
Comment (adelowo) [inline models/repo.go:1120]: Should in case anyone is looking at this, it was moved to the newly created `repo_transfer.go` file https://github.com/go-gitea/gitea/pull/5744/files#diff-4a00b23928d7e63417b6a39f3052ec23R126

File: models/migrations/v78.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package migrations
```
Comment (lunny) [inline models/migrations/v78.go]: comment header

File: models/migrations/v78.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"code.gitea.io/gitea/modules/util"
+	"github.com/go-xorm/xorm"
```
Comment (lunny) [inline models/migrations/v78.go]: blank line

File: models/migrations/v78.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+func addRepoTransfer(x *xorm.Engine) error {
+	type RepoTransfer struct {
+		ID          int64 `xorm:"pk autoincr"`
+		RecipientID int64
```
Comment (lunny) [inline models/migrations/v78.go]: unique index between `recipient_id` and `repo_id`.

File: models/repo_transfer.go [hunk 5]
```diff
@@ -0,0 +1,237 @@
+package models
```
Comment (lunny) [inline models/repo_transfer.go:5]: comment header

File: models/repo_transfer.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+import (
+	"fmt"
+	"os"
+
+	"code.gitea.io/gitea/modules/util"
+	"github.com/Unknwon/com"
+)
+
+// RepoStatus determines the current state of a transfer
+type RepoStatus uint8
```
Comment (lunny) [inline models/repo_transfer.go]: Maybe name it `TransferStatus` better?

File: models/repo_transfer.go [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+const (
+	// Rejected is a status for transfers that get cancelled by either the
+	// recipient or the user who initiated the transfer
+	Rejected RepoStatus = iota
```
Comment (lunny) [inline models/repo_transfer.go]: Why not `Pending` is the default value?

File: models/repo_transfer.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+
+	transfer := &RepoTransfer{
+		RepoID:      repo.ID,
+		RecipientID: newOwner.ID,
+		Status:      Pending,
+		CreatedUnix: util.TimeStampNow(),
+		UpdatedUnix: util.TimeStampNow(),
+	}
+
+	_, err = x.Insert(transfer)
```
Comment (lunny) [inline models/repo_transfer.go:193]: And can user still push codes or something on this repository?

**Review by lunny — CHANGES_REQUESTED** [review]

File: models/repo_transfer.go (hunk 7 above)
Comment (adelowo) [reply models/repo_transfer.go]: When writing to the database, Pending is the default. 

I wanted 

Rejected - 0
Pending - 1
Accepted - 2 

as it is mentally easier to map 0 to false ( Rejected)

File: models/repo_transfer.go (hunk 8 above)
Comment (adelowo) [reply models/repo_transfer.go:193]: I will still have to look into how Github does this but my instinct says pushing should still be allowed

File: models/repo_transfer.go (hunk 7 above)
Comment (Larivact) [reply models/repo_transfer.go]: With states it is common practice to have 0 be the default. Rejected is not false, it's a state.

**Comment by adelowo** [comment]
So I have this guy already working fine. I just need to send an email to the potential new owner.

Thinking out loud if it makes any sense creating a redirect for the repo after a transfer

**Comment by adelowo** [comment]
Will try to wrap up the email and tests this weekend

File: models/repo_transfer.go (hunk 8 above)
Comment (adelowo) [reply models/repo_transfer.go:193]: Yes, code can still go into the repository..

**Comment by techknowlogick** [comment]
Posting this here in case it gets lost on discord: I think `RepoTransfer` needs to be added to this list: https://github.com/go-gitea/gitea/blob/91dcccf72d15d9d6dbe9c0cc5c5d33c082225cde/models/models.go#L75

**Comment by adelowo** [comment]
@techknowlogick can the `status/wip` label be removed now ? This is ready for review :)

File: routers/repo/setting.go [hunk 9]
```diff
@@ -313,43 +314,83 @@ func SettingsPost(ctx *context.Context, form auth.RepoSettingForm) {
 
 	case "transfer":
 		if !ctx.Repo.IsOwner() {
-			ctx.Error(404)
+			ctx.Error(http.StatusUnauthorized)
```
Comment (lafriks) [inline routers/repo/setting.go]: It should probably be 404 at least for users not having read access to repositiry

**Comment by lafriks** [comment]
Where does recipient sees that he has repository transfer to accept? (In case no mail notifications are enabled)?

**Comment by adelowo** [comment]
Ha true... I didn’t think of that usecase before .

What do you suggest ? I will think about this throughout the night

**Comment by jolheiser** [comment]
What about a new notification, regardless of email?

**Comment by adelowo** [comment]
Fat fingers :)

**Comment by adelowo** [comment]
@jolheiser I went with your suggestion, it appears as a notification now. Can you test it when you do have time?

cc @lafriks

**Comment by adelowo** [comment]
^^^

**Comment by lafriks** [comment]
Conflicts ;)

**Comment by adelowo** [comment]
Ha, I was on mobile. Didn’t see those. Will resolve

**Comment by adelowo** [comment]
Have to revive this for 1.12

**Comment by 6543** [comment]
the logick you added at models/repo_transfer.go
should go in `services/repository/transfer.go` <- there are the notification things now ...

and if you the aim is a organisation you can create repos at or if you are admin it should migrate as normal

**Comment by adelowo** [comment]
Yeah, I saw it.. Lots of merge conflicts. Will fix after work tomorrow

File: go.mod [hunk 10]
```diff
@@ -45,6 +46,7 @@ require (
 	github.com/go-redis/redis v6.15.2+incompatible
 	github.com/go-sql-driver/mysql v1.4.1
 	github.com/go-swagger/go-swagger v0.21.0
+	github.com/go-xorm/xorm v0.7.9
```
Comment (6543) [inline go.mod]: why do you use a old xorm ?

-> xorm.io/xorm v0.8.2-0.20200120024500-c37aff9b3a4a

File: models/migrations/v126.go [hunk 11]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/modules/util"
+
+	"github.com/go-xorm/xorm"
```
Comment (6543) [inline models/migrations/v126.go]: old xorm version!!

File: go.mod (hunk 10 above)
Comment (adelowo) [reply go.mod]: Merge issues.. Will resolve

**Comment by lafriks** [comment]
Please fix lint errors https://drone.gitea.io/go-gitea/gitea/20557/1/5

Also unit test for this PR seems to fail

**Comment by adelowo** [comment]
I just merged the conflicts. The actual implementation is broken. Will fix

**Comment by lafriks** [comment]
Tests fail with compilation error:
```
integrations/api_repo_test.go:437:4: unknown field 'TeamIDs' in struct literal of type structs.TransferRepoOption
```

File: modules/structs/repo.go [hunk 12]
```diff
@@ -163,8 +163,6 @@ type EditRepoOption struct {
 type TransferRepoOption struct {
 	// required: true
 	NewOwner string `json:"new_owner"`
-	// ID of the team or teams to add to the repository. Teams can only be added to organization-owned repositories.
```
Comment (6543) [inline modules/structs/repo.go]: why do you remove this function?

Comment (6543) [reply modules/structs/repo.go]: if the actor is admin of the transfering repo AND the aim organisation he can asign Teams to it...

File: integrations/api_repo_test.go [hunk 13]
```diff
@@ -434,7 +434,6 @@ func TestAPIRepoTransfer(t *testing.T) {
 		token = getTokenForLoggedInUser(t, session)
 		req = NewRequestWithJSON(t, "POST", fmt.Sprintf("/api/v1/repos/%s/%s/transfer?token=%s", repo.OwnerName, repo.Name, token), &api.TransferRepoOption{
 			NewOwner: testCase.newOwner,
-			TeamIDs:  testCase.teams,
```
Comment (6543) [inline integrations/api_repo_test.go]: why remove this testcase ?

**Review by 6543 — CHANGES_REQUESTED** [review]
why remove a function?

File: modules/structs/repo.go (hunk 12 above)
Comment (adelowo) [reply modules/structs/repo.go]: Mapping the API behavior to the regular transfer via web

Comment (6543) [reply modules/structs/repo.go]: @adelowo -> https://developer.github.com/v3/repos/#transfer-a-repository is having this too ...

Comment (6543) [reply modules/structs/repo.go]: I think we can support repo-confirm & teamID ..
* If actor has right to create a repo and manage it on destination user (becauseo of admin rights or owner of an organisation) -> take kare of TeamID
* If actor has NO right to destination -> Ignore TeamID (perhaps a warning to log)

**Comment by lafriks** [comment]
Swagger needs to be regenerated

**Comment by adelowo** [comment]
@lunny @6543 @lafriks ready for review

File: models/migrations/v128.go [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+func addRepoTransfer(x *xorm.Engine) error {
+	type RepoTransfer struct {
+		ID          int64 `xorm:"pk autoincr"`
+		UserID      int64
+		RecipientID int64
+		RepoID      int64
+		CreatedUnix timeutil.TimeStamp `xorm:"INDEX NOT NULL created"`
+		UpdatedUnix timeutil.TimeStamp `xorm:"INDEX NOT NULL updated"`
+		Status      bool
```
Comment (6543) [inline models/migrations/v128.go]: Please update this struct ot use models.RepoTransferbdirect

Comment (adelowo) [reply models/migrations/v128.go]: Nice catch... Fixed

**Comment by adelowo** [comment]
Unrelated test failure.

**Comment by adelowo** [comment]
`TestGit/HTTP/MergeFork` fails randomly.

**Comment by adelowo** [comment]
Unrelated failure : `TestGit/HTTP/MergeFor`

**Comment by 6543** [comment]
tested this pull: resiver get no notivication in ui

**Comment by 6543** [comment]
and secound test didn't worked:
1. Create Repo `repA`
2. Create Org `orgA`
3. Transfer repA to orgA
-> repo is in pending state :( -> why I have to acept a repo transver to a place i have admin access to?

**Comment by adelowo** [comment]
Hmm, I should get rid of the UI notification. It’s not needed actually. Made a mental note to remove it but forgot. There’s no sense to it being in the notification or so I think. GitHub doesn’t do that either.

Secondly, there’s a nice case. Will handle it

File: modules/notification/ui/ui.go [hunk 15]
```diff
@@ -26,7 +26,7 @@ type (
 )
 
 var (
-	_ base.Notifier = &notificationService{}
+	_ base.Notifier = (*notificationService)(nil)
```
Comment (6543) [inline modules/notification/ui/ui.go:29]: unrelated change

**Comment by 6543** [comment]
pleace resolve conflics :O

**Comment by adelowo** [comment]
Resolved

**Comment by adelowo** [comment]
> Hmm, I should get rid of the UI notification. It’s not needed actually. Made a mental note to remove it but forgot. There’s no sense to it being in the notification or so I think. GitHub doesn’t do that either.
> 
> Secondly, there’s a nice case. Will handle it

Will get to this over the weekend

**Comment by adelowo** [comment]
2 years ago.. Damn, I should pick this up again.

**Comment by lunny** [comment]
- [ ] Transfer to an organization that you have permisstion to create repository, the confirmation should be ignored.
- [ ] If the limitation is max for the receiver, it should be failed.
- [ ] When it's in transfering state, the repository should be in read-only state.

**Comment by resynth1943** [comment]
Please merge this immediately.

The current behaviour makes it look like you have been hacked, which is actually quite terrifying.
I woke up to this today, and assumed my account was compromised.

The best way to move forward would be a user interface element (like a notification) that asks
users whether or not they wish to take on these new repositories.

**Comment by techknowlogick** [comment]
@resynth1943 we are all volunteers working on this for free. If you’d like development to go at a pace you prefer you are free to resolve the issues above yourself or pay someone to do it for you.

**Comment by resynth1943** [comment]
@techknowlogick not sure what your problem is, mate. Simply offering a use-case in which this can be abused, and a possible way of implementing it.

**Comment by lafriks** [comment]
@resynth1943 it can not be merged while it is not finished so you can't really request to merge it immediately, in fact requesting someone to do for free anything is very bad tone

**Comment by resynth1943** [comment]
Thanks for sharing. I'd like to focus on the pull request, though ❤️ 

Is there a mockup of what the UI looks like in this new rendition?

**Comment by 6543** [comment]
just an idear: but we could extend RepoState and add "RepoTransvering" next to Migrating status ...

**Comment by 6543** [comment]
I'll fix and finish this one or open a new pull based on this :)

**Comment by 6543** [comment]
close since #14792 is here

**Comment by 6543** [comment]
@resynth1943 #14792 has a demo video now ...
